### ISSUE TITLE
Raku -v has blank characters when Unicode not available

### ISSUE THREAD

**Issue by xsawyerx** [rakudo/rakudo#4505]
## The Problem

I can't see the pretty Unicode "raku" name when running `raku -v`

## Expected Behavior

```
Welcome to 𝐑𝐚𝐤𝐮𝐝𝐨™ v2021.08.
Implementing the 𝐑𝐚𝐤𝐮™ programming language v6.d.
Built on MoarVM version 2021.08.
```

## Actual Behavior

```
Welcome to [][][][][][]™ v2021.08.
Implementing the [][][][]™ programming language v6.d.
Built on MoarVM version 2021.08.
```

## Steps to Reproduce

1. Using `urxvt`
2. Configuration in `~/.Xdefaults`:

```
URxvt*font: xft:SourceCodeVariable:size=10
URxvt*boldFont: xft:SourceCodeVariable:bold:size=10
```

3. Run `raku -v`
4. Be sad

## Environment

* Ubuntu 21.04
* Fresh install of `rakubrew`
* `rxvt-unicode` version 9.22-8build1
* `raku` version v2021.08

**Comment by lizmat** [comment]
Thanks for the report!

Quick answer is: get Unicode support  :-)

Also, this is a known issue  :-(

Longer answer: once the trademark application for Raku and Rakudo is accepted, we will probably change it back to all ASCII for all operating systems.

**Comment by lizmat** [comment]
No, we test for Windows.  :-)

**Comment by vrurg** [comment]
Perhaps, it makes sense to? Somehow I've got used to it and will miss that bold 𝐑𝐚𝐤𝐮 if it's gone. :)

**Comment by lizmat** [comment]
Well, tell me how to look for that setting in an OS independent way... and I'll put it in  :-)

**Comment by vrurg** [comment]
I think we can start with testing for `%*ENV<LC_CTYPE> // %*ENV<LANG>` to contain `\.UTF` in them for any other OS but Windows. Should eliminate most of related problems, AFAIK.

**Comment by pmichaud** [comment]
> Thanks for the report!
> 
> Quick answer is: get Unicode support :-)

Nominally, the original poster is using a system that has "Unicode support".  Like the OP, I also use urxvt as my terminal of choice (specifically for its stated Unicode support), and I also see the rectangles instead of "Rakudo"/"Raku" when I invoke rakudo from the command line.  So saying "get Unicode support" is really not helpful as an answer to me.

So, given that I'm using a terminal program that claims to have Unicode support, I'm still not seeing what I expect to see here.  I spent a few minutes trying to chase this down on my system, and it likely involves properly setting the LC_TYPE environment variable prior to invoking my terminal emulator, but I tried that and it still didn't resolve the issue....

All of which just begs the point that having rectangles show up for the sake of "pretty font" really violates the principle of least surprise.  If I were a new user to Rakudo/Raku, having this as my initial experience would be a negative for me.  That alone argues to me that it'd be better to just use ASCII rather than playing a "oh, your system's apparently not worthy of Raku" sort of card.

Pm

**Comment by pmichaud** [comment]
> I think we can start with testing for `%*ENV<LC_CTYPE> // %*ENV<LANG>` to contain `\.UTF` in them for any
> other OS but Windows. Should eliminate most of related problems, AFAIK.

As noted in my response just posted above, having LC_CTYPE set to "en_US.UTF_8" (my default setting) doesn't seem to resolve the issue in urxvt.   It's fine with me if we just claim this is a urxvt bug and it's actually not displaying Unicode properly, but simply checking for LC_CTYPE for /UTF/ does not appear to be sufficient in my case (and perhaps not the OP's case either).

Pm

**Comment by lizmat** [comment]
To be clear once again: the emphasis in font was deemed necessary to be able to get a trademark application passed.  Yes, I know, not the best of reasons  :-(

**Comment by pmichaud** [comment]
To be clear:  the original answer didn't say "we have it this way in order to support a trademark application".  :) The original answer simply says it'll probably revert to ASCII once trademark is accepted, which read to me at the time like a bit of a non-sequitur.  Now that the relationship to trademark is clear, it makes more sense.  :)  

Pm

**Comment by vrurg** [comment]
@pmichaud if setting `LANG` alone is insufficient to get urxvt to work (BTW, hopefullt it's using a Unicode font?), then I would only be in favor of getting rid of the Unicode in the version output. After all, there are sorts of things that must not require special skills to make them work.

**Comment by pmichaud** [comment]
@vrurg   Yes, I checked and I think I'm using a unicode font  ("Hack" on ubuntu) in urxvt.  When I run rakudo from a Konsole window, I do see the "rakudo"/"raku" correctly, and as far as I can tell Konsole is using the same "Hack" font.

Pm

**Comment by xsawyerx** [comment]
Thank you for all the replies. :)

I had indeed forgotten to add the locale definitions, so:

```
$ locale
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=en_US.UTF-8
LC_TIME=en_US.UTF-8
LC_COLLATE="en_US.UTF-8"
LC_MONETARY=en_US.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=en_US.UTF-8
LC_NAME=en_US.UTF-8
LC_ADDRESS=en_US.UTF-8
LC_TELEPHONE=en_US.UTF-8
LC_MEASUREMENT=en_US.UTF-8
LC_IDENTIFICATION=en_US.UTF-8
LC_ALL=
```

(I also updated the description in the ticket.)

So I have things set to UTF-8 and I have a Unicode terminal. Perhaps the font doesn't support Unicode.

**Comment by jonathanstowe** [comment]
It seems to support _some_ Unicode because it is showing the "™", which is odd.  But I can confirm that it does the same with `urxvt` here too, the boxes rather than mojibake would also seem to suggest fontiness too.

**Comment by jonathanstowe** [comment]
I'm beginning to wonder whether this particular case is a bug in the font selection in urxvt because similarly if I supply it with the same font list that my regular terminal uses it still does the same.

**Comment by patrickbkr** [comment]
I'm a strong supporter of fixing this in a robust way, sacrificing shinyness if necessary. This issue has been reported very regularly.

Is there a timeline for the trademark application? If it's long off, I'd propose searching for an alternate solution.
All caps? Underlining? ASCII art?

**Comment by niner** [comment]
This looks, sounds and smells like a font issue. So why is it possible that the same font works e.g. in konsole but not in urxvt?

Just an educated guess: I could imagine that these bold characters are not just plain characters in the font but use some mechanism that has to be supported by the program using the font to render them bold. Remember that true type fonts are actually executable code. So if urxvt doesn't support this mechanism, it may still fall back to render the missing glyph symbol (i.e. that square).
